realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java

```diff
@@ -0,0 +1,511 @@
+     * @param fieldName existing field name to check.
+     * @return {@code true} if field is indexed, {@code false} otherwise.
+     * @see io.realm.annotations.Index
+     */
+    public boolean hasIndex(String fieldName) {
+        checkLegalName(fieldName);
+        checkFieldExists(fieldName);
+        return table.hasSearchIndex(table.getColumnIndex(fieldName));
+    }
+
+
+    /**
+     * Removes an index from a given field. This is the same as removing the {@code @Index} annotation on the field.
+     *
+     * @param fieldName field to remove index from.
+     * @return the updated schema.
+     * @throws IllegalArgumentException if field name doesn't exists or the field doesn't have an index.
+     */
+    public RealmObjectSchema removeIndex(String fieldName) {
+        checkLegalName(fieldName);
+        checkFieldExists(fieldName);
+        long columnIndex = getColumnIndex(fieldName);
+        if (!table.hasSearchIndex(columnIndex)) {
+            throw new IllegalStateException("Field is not indexed: " + fieldName);
+        }
+        table.removeSearchIndex(columnIndex);
+        return this;
+    }
+
+    /**
+     * Adds a primary key to a given field. This is the same as adding the {@link io.realm.annotations.PrimaryKey}
+     * annotation on the field.
+     *
+     * @param fieldName field to add name to rename.
+     * @return the updated schema.
+     * @throws IllegalArgumentException if field name doesn't exists, the field cannot be a primary key or it already
+     * has a primary key defined.
+     */
+    public RealmObjectSchema addPrimaryKey(String fieldName) {
+        checkLegalName(fieldName);
+        checkFieldExists(fieldName);
+        if (table.hasPrimaryKey()) {
+            throw new IllegalStateException("A primary key is already defined");
+        }
+        table.setPrimaryKey(fieldName);
+        return this;
+    }
+
+    /**
+     * Removes the primary key from this class. This is the same as removing the {@link io.realm.annotations.PrimaryKey}
+     * annotation from the class.
+     *
+     * @return the updated schema.
+     * @throws IllegalArgumentException if the class doesn't have a primary key defined.
+     */
+    public RealmObjectSchema removePrimaryKey() {
+        if (!table.hasPrimaryKey()) {
+            throw new IllegalStateException(getClassName() + " doesn't have a primary key.");
+        }
+        table.setPrimaryKey("");
+        return this;
+    }
+
+    /**
+     * Sets a field to be required, i.e. not allowed to hold {@code null values}. This is equivalent to switching
+     * between boxed types and their primitive variant e.g. {@code Integer} to {@code int}.
+     *
+     * @param fieldName name of field in the class.
+     * @param required  {@code true} if field should be required, {@code false} otherwise.
+     * @return the updated schema.
+     * @throws IllegalArgumentException if field name doesn't exists or the field already has the given required flag.
```

Comment:
if field name doesn't exists, cannot be applied with @{@link Required} or the field already has the given required flag.